components/plurals/src/rules/runtime/ast.rs

```diff
@@ -0,0 +1,634 @@
+///  * One bit to store the state of AndOr.
+///  * One bit to store the state of Polarity.
+///  * Six bits to store the Operand.
+///
+/// The following four bytes store the value of the modulo as `u32`.
+///
+/// Finally, the remaining variable length is used to store a list
+/// of `RangeOrValue` elements which have `[u32;2]` layout.
+///
+/// # Diagram
+///
+/// ```text
+/// ┌───────────────┬───────────────┬───────────────┐
+/// │      u8       │    [u8;4]     │     [u8]      │
+/// ├─┬─┬─┬─┬─┬─┬─┬─┼───┬───┬───┬───┼───────────────┤
+/// ├─┼─┼─┴─┴─┴─┴─┴─┼───┴───┴───┴───┼───────────────┤
+/// │ │ │  Operand  │    Modulo     │  RangeOrValue │
+/// └─┴─┴───────────┴───────────────┴───────────────┘
+///  ▲ ▲
+///  │ │
+///  │ Polarity
+///  │
+///  AndOr
+/// ```
+///
+/// # Optimization
+///
+/// This model is optimized for efficient packaging of the `Relation` elements
+/// and performant deserialization from the `RelationULE` to `Relation` type.
+///
+/// # Constraints
+///
+/// The model constraints the `Operand` to 64 variants, and `Modulo` to `u32::MAX`.
+#[derive(Debug)]
+#[repr(packed)]
+pub struct RelationULE {
+    /// This maps to (AndOr, Polarity, Operand),
+    /// with the first bit mapping to AndOr (1 == And), the second bit
+    /// to Polarity (1 == Positive), and the remaining bits to Operand
+    /// encoded via Operand::encode. It is unsound for the Operand bits to
+    /// not be a valid encoded Operand.
+    andor_polarity_operand: u8,
+    modulo: <u32 as AsULE>::ULE,
+    range_list: [RangeOrValueULE],
+}
+
+impl RelationULE {
+    pub fn as_relation(&self) -> Relation {
+        // safe to call because we are operating on a valid RelationULE
+        let (and_or, polarity, operand) =
+            unsafe { Self::decode_andor_polarity_operand(self.andor_polarity_operand) };
+        Relation {
+            and_or,
+            polarity,
+            operand,
+            modulo: u32::from_unaligned(self.modulo),
+            range_list: ZeroVec::borrowed_from_slice(&self.range_list),
+        }
+    }
+
+    fn encode_andor_polarity_operand(and_or: AndOr, polarity: Polarity, operand: Operand) -> u8 {
+        // XXX: Ensure and_or is one bit, polarity is one bit, and operand is max 6 bits
+        (((and_or == AndOr::And) as u8) << 7)
+            + (((polarity == Polarity::Positive) as u8) << 6)
+            + operand.idx()
+    }
+
+    // Must be called on an `encoded` that is obtained from `encode_andor_polarity_operand`, i.e.
+    // the field on a valid RelationULE
+    unsafe fn decode_andor_polarity_operand(encoded: u8) -> (AndOr, Polarity, Operand) {
+        let and_or = if encoded & 0b1000_0000 != 0 {
+            AndOr::And
+        } else {
+            AndOr::Or
+        };
+
+        let polarity = if encoded & 0b0100_0000 != 0 {
+            Polarity::Positive
+        } else {
+            Polarity::Negative
+        };
+
+        let operand = Operand::from_idx(encoded & 0b0011_1111);
```

Comment:
See my updated comment above regarding `Operand::from_idx`.  It is correct here to use the infallible (panicking or unsafe) version of that function, since we are inside an `unsafe` function.